contract_compassion/tests/test_contract_compassion.py

```diff
@@ -0,0 +1,339 @@
+        invoices = invoicer.invoice_ids
+        nb_invoices = len(invoices)
+        self.assertEqual(nb_invoices, 6)
+        invoice3 = invoices[3]
+        self.assertEqual(invoice3.state, 'open')
+        # Payment of the third invoice so the
+        # contract will be on the active state and the 2 first invoices should
+        # be canceled.
+        self._pay_invoice(invoice3.id)
+        invoice_paid = invoice_obj.browse(self.cr, self.uid, invoices[0].id)
+        invoice_paid1 = invoice_obj.browse(self.cr, self.uid, invoices[1].id)
+        invoice_paid2 = invoice_obj.browse(self.cr, self.uid, invoices[2].id)
+        invoice_paid3 = invoice_obj.browse(self.cr, self.uid, invoice3.id)
+        invoice_paid4 = invoice_obj.browse(self.cr, self.uid, invoices[4].id)
+        invoice_paid5 = invoice_obj.browse(self.cr, self.uid, invoices[5].id)
+        contract_act = contract_obj.browse(self.cr, self.uid, self.contract_id)
+        self.assertEqual(invoice_paid3.state, 'paid')
+        self.assertEqual(invoice_paid.state, 'open')
+        self.assertEqual(invoice_paid1.state, 'open')
+        self.assertEqual(invoice_paid2.state, 'open')
+        self.assertEqual(invoice_paid4.state, 'cancel')
+        self.assertEqual(invoice_paid5.state, 'cancel')
+        self.assertEqual(contract_act.state, 'active')
+        wf_service.trg_validate(
+            self.uid, 'recurring.contract',
+            self.contract_id, 'contract_terminated', self.cr)
+        contract_term = contract_obj.browse(
+            self.cr, self.uid, self.contract_id)
+        self.assertEqual(contract_term.state, 'terminated')
+
+    def test_contract_compassion_second_scenario(self):
+        """
+            Testing if invoices are well concelled when we cancel the related
+            contract.
+        """
+        # Testing area creation
+        self.contract_group = self._create_group(
+            'do_nothing', 1, 'month', self.partner_id1, 1,
+            self.payment_term_id)
+        self.contract_id = self._create_contract(
+            datetime.today().strftime(DF), self.contract_group, 'postal',
+            datetime.today().strftime(DF))
+        self.contract_line_id = self._create_contract_line(
+            self.contract_id, '3', '200.0')
+
+        # Switch to "waiting for payment" state
+        wf_service = netsvc.LocalService('workflow')
+        wf_service.trg_validate(
+            self.uid, 'recurring.contract',
+            self.contract_id, 'contract_validated', self.cr)
+        contract_obj = self.registry('recurring.contract')
+        contract_wait = contract_obj.browse(
+            self.cr, self.uid, self.contract_id)
+        invoicer_obj = self.registry('recurring.invoicer')
+        invoicer_id = contract_wait.button_generate_invoices()
+        invoicer = invoicer_obj.browse(self.cr, self.uid, invoicer_id)
+        invoices = invoicer.invoice_ids
+        invoice = invoices[0]
+        invoice1 = invoices[1]

```

Comment:
Again, you don't need all these variables. Try to make the code lighter. You can just have :
```
invoices = invoicer_obj.browse(self.cr, self.uid, invoicer_id).invoice_ids
assert(invoices[0], ...)
assert(invoices[1], ...)
```